packages/react-storage/src/components/StorageBrowser/actions/handlers/listLocations.ts

```diff
@@ -1,21 +1,85 @@
-import { Permission } from '../../storage-internal';
+import {
+  ListLocationsOutput,
+  Permission,
+  listCallerAccessGrants,
+} from '../../storage-internal';
+import { assertListCallerAccessGrantInput } from '../../validators';
+
+import {
+  ListHandlerOptions,
+  ListHandlerInput,
+  ListHandlerOutput,
+  ListHandler,
+} from '../types';
 
-import { ListHandlerOptions, ListHandlerOutput, ListHandler } from '../types';
 import { LocationData, LocationType } from './types';
+import { parseLocations } from './utils';
+
+const DEFAULT_PAGE_SIZE = 1000;
 
 type ExcludeType = Permission | LocationType;
 
 export interface ListLocationsHandlerOptions
   extends ListHandlerOptions<ExcludeType | ExcludeType[]> {}
 
-export interface ListLocationsHandlerInput {
-  options?: ListLocationsHandlerOptions;
-}
+export interface ListLocationsHandlerInput
+  extends ListHandlerInput<ListLocationsHandlerOptions> {}
+
 export interface ListLocationsHandlerOutput
   extends ListHandlerOutput<LocationData> {}
 
 export interface ListLocationsHandler
   extends ListHandler<ListLocationsHandlerInput, ListLocationsHandlerOutput> {}
 
-export const listLocationsHandler: ListLocationsHandler =
-  null as unknown as ListLocationsHandler;
+const shouldExclude = (
+  permission: Permission,
+  type: LocationType,
+  exclude?: ExcludeType | ExcludeType[]
+) =>
+  !exclude
+    ? false
+    : typeof exclude === 'string'
+    ? exclude === permission || exclude === type
+    : exclude.includes(permission) || exclude.includes(type);
+
+export const listLocationsHandler: ListLocationsHandler = async (input) => {
+  const { config, options } = input;
+  const { accountId, credentials, region } = config;
+  const { exclude, nextToken, pageSize = DEFAULT_PAGE_SIZE } = options ?? {};
+
+  const fetchLocations = async (
+    accumulatedItems: LocationData[],
+    locationsNextToken: ListLocationsOutput['nextToken']
+  ): Promise<{
+    items: LocationData[];
+    nextToken: ListLocationsOutput['nextToken'];
+  }> => {
+    const remainingPageSize = pageSize - accumulatedItems.length;
+
+    if (!accountId) {
```

Comment:
is `!accountId` redundant here for the assertion below?

Reply:
Yes upadted